### PR TITLE
Apache configuration improvements

### PR DESCRIPTION
**Pull request by wmertens**
- Verify configuration with httpd -t
- Some minor editing of option descriptions
- Make Event MPM the default on v2.4 

@edolstra any objections? The config check can be turned off and Event is good starting with kernel 2.6...

### CODE REVIEW

**Comment by lethalman** [comment]
Did you try PHP with mpm? It's a different setup, I guess this is going to break every php deployment.

**Comment by wmertens** [comment]
Hmm I see, mod_php is threadsafe nowadays but extensions may not be... http://serverfault.com/questions/383526/how-do-i-select-which-apache-mpm-to-use

So I propose we force the event model to be prefork when mod_php is enabled?

File: nixos/modules/services/web-servers/apache-httpd/default.nix [hunk 1]
```diff
@@ -417,6 +419,19 @@ let
...
+  verifiedConf = pkgs.runCommand "apache.conf" {} ''
+    c=${mainCfg.configFile}
+    trap "echo '==> configuration error, see $c'" ERR
+
+    # Allow testing in sandbox
+    sed "s:/run/httpd:$PWD:" < $c > test-conf.conf
+    mkdir runtime
+
+    echo "verifying Apache configuration, ignore warnings about missing directories"
+    ${httpd}/bin/httpd -f $PWD/test-conf.conf -t
```
Comment (edolstra) [inline nixos/modules/services/web-servers/apache-httpd/default.nix]: Does this work if the Apache user/group don't exist?

File: nixos/modules/services/web-servers/apache-httpd/default.nix [hunk 2]
```diff
@@ -613,7 +628,13 @@ in
       enableCompression = mkOption {
         type = types.bool;
         default = false;
-        description = "Enable compression of responses using mod_deflate.";
+        description = "Enable compression of responses using <literal>mod_deflate</literal>.";
+      };
+
+      verifyConfig = mkOption {
```
Comment (edolstra) [inline nixos/modules/services/web-servers/apache-httpd/default.nix]: Why make this optional?

**Comment by lethalman** [comment]
@wmertens no it's just completely different. mpm event is similar to how you setup nginx with php using fpm and external servers. It doesn't use mod_php.

**Comment by wmertens** [comment]
@lethalman if one is not using mod_php there is no problem, that works fine
with event and worker…

On Tue, Aug 25, 2015, 10:43 lethalman <notifications@github.com> wrote:

> @wmertens <https://github.com/wmertens> no it's just completely
> different. mpm event is similar to how you setup nginx with php using fpm
> and external servers. It doesn't use mod_php.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/9435#issuecomment-134524953>.
>
-- 

Wout.
(typed on mobile, excuse terseness)

File: nixos/modules/services/web-servers/apache-httpd/default.nix (hunk 2 above)
Comment (wmertens) [inline nixos/modules/services/web-servers/apache-httpd/default.nix]: I was reasoning that there might be edge cases where the test fails during build and wanted to provide an out.

**Comment by wmertens** [comment]
Here's how PHP should be configured on Apache v2.4: http://wiki.apache.org/httpd/PHP-FPM

I think enabling PHP on the service should configure that instead of mod_php.

**Comment by wmertens** [comment]
I just realized that choosing the MPM module based on the version rebuilds the man pages when changing the package version so I took that out again.

@edolstra the test now works when the user doesn't exist yet, good catch :)

**Comment by globin** [comment]
Yeah we should definitely go for using php-fpm only and ditch mod_php if possible. PHP internals has thought about deprecating mod_php but it sadly still is used too often to get rid of it.. I'll have a look at freshening up the fpm module to be configurable in a more human-friendly way soon and try to get support for that into apache by flicking over a simple switch.

**Comment by edolstra** [comment]
Note that we used to check the Apache configuration file, but we got rid of it because it was too much trouble: 4a95f8996b11ca9876e24a37d4b521925a1c33bf

**Comment by wmertens** [comment]
Another reason for having an out… looks like I need to fix ssl key testing
in sandbox :-)

On Wed, Aug 26, 2015, 11:00 Eelco Dolstra <notifications@github.com> wrote:

> Note that we used to check the Apache configuration file, but we got rid
> of it because it was too much trouble: 4a95f89
> <https://github.com/NixOS/nixpkgs/commit/4a95f8996b11ca9876e24a37d4b521925a1c33bf>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/9435#issuecomment-134905047>.
>
-- 

Wout.
(typed on mobile, excuse terseness)

**Comment by rasendubi** [comment]
(triage) @wmertens status?

**Comment by wmertens** [comment]
Hmm totally forgot about this one. IIRC the problem was that when testing the configuration it needed to access the network for verifying SSL certificates. I suppose that could be fixed by stripping them out before testing, or at least filter out errors due to SSL certificates.

Problem is that I'm not using Apache at the moment. Anybody up for completing this?

**Comment by wmertens** [comment]
It is indeed too bothersome to have this.
